### PR TITLE
Initial contribution of property-view extension

### PR DESCRIPTION
**Pull request by ndoschek**
#### Introduction

Many IDEs, such as the traditional Eclipse IDE, have the notion of a global, extensible property view, which shows additional information about the current selection within the IDE. Property views are heavily used in those IDEs for showing details of elements in e.g. diagram editors, complex tree editors, or the file explorer selection. Therefore, the main idea is to have a global, generic property view in the IDE, but allow specific implementations to extend the contents of the global property view with specific additional information for some type of selection.

#### What it does

The `@theia/property-view` extension contributes a generic, global property view based on Theia's global selection.
The property view widget can be opened/toggled either via menu `View->Properties` or via shortcut `Shift+Alt+P`. It is located in the bottom dock area per default.
The following two default content widgets are implemented in this extension:
- EmptyPropertyViewWidget: If no other widget can be provided, a simple message (No properties available) is shown.
- ResourcePropertyViewWidget: Can handle file and monaco editor selections and displays the general FileStat information (e.g. location, name, last modified) in a TreeWidget.

To contribute a specific property view, it is necessary to implement a `PropertyViewDataService` which gathers the property data for a selection as well as a `PropertyViewWidgetProvider` which provides a suitable content widget to display the property data for a specific selection inside the property view widget.

#### How to test

- Build Theia and open either the browser or electron example.
- Open any workspace (which preferably is holding directories and files, otherwise please add a few)
- Open the property view either via menu `View -> Properties` or via Shortcut `Shift+Alt+P`
- Select a file or directory and view its properties in a tree widget structure
- Other selections (or deselections of files/directories) should result in an empty property view (`No properties available.`)
- Tooltip is available for the property items

#### Review checklist

- [X] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
 
#### Example implementation

Here is a short example of how to implement an additional property view, which displays the name and whether it is a file or directory in a simple React widget based on the selection from the file explorer (assuming there would be no `ResourcePropertyViewWidget` of course):

The `FileInfoPropertyDataService` gathers the file information and delivers a custom object:
```ts
export interface FileInfoPropertyObject {
    name: string;
    isDirectory: boolean;
}
 
@injectable()
export class FileInfoPropertyDataService implements PropertyDataService {
 
    readonly id = 'fileinfo';
    readonly label = 'FileInfoPropertyDataService';
 
    @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
 
    canHandleSelection(selection: Object | undefined): number {
        return this.isFileSelection(selection) ? 1 : 0;
    }
 
    private isFileSelection(selection: Object | undefined): boolean {
        return !!selection && Array.isArray(selection) && FileSelection.is(selection[0]);
    }
 
    async providePropertyData(selection: Object | undefined): Promise<FileInfoPropertyObject | undefined> {
        if (this.isFileSelection(selection) && Array.isArray(selection)) {
            return {
                name: this.labelProvider.getName(selection[0].fileStat.resource),
                isDirectory: (selection[0].fileStat as FileStat).isDirectory
            };
        }
        return Promise.reject();
    }
}
```
 
The `FileInfoPropertyWidget` is a simple ReactWidget and displays the selected node and whether it is a file or directory:
```ts
class FileInfoPropertyViewWidget extends ReactWidget implements PropertyViewContentWidget {
 
    static readonly ID = 'file-info-property-view';
    static readonly LABEL = 'File Information';
 
    protected currentFileInfo: FileInfoPropertyObject;
 
    constructor() {
        super();
        this.id = FileInfoPropertyViewWidget.ID;
        this.title.label = FileInfoPropertyViewWidget.LABEL;
        this.title.caption = FileInfoPropertyViewWidget.LABEL;
        this.title.closable = false;
        this.node.tabIndex = 0;
    }
 
    updatePropertyViewContent(propertyDataService?: PropertyDataService, selection?: Object | undefined): void {
        if (propertyDataService) {
            propertyDataService.providePropertyData(selection).then((fileInfo: FileInfoPropertyObject) => this.currentFileInfo = fileInfo);
        }
        this.update();
    }
 
    protected render(): React.ReactNode {
        return <div>
            {`Selected node in explorer: ${this.currentFileInfo.name} ${this.currentFileInfo.isDirectory ? '(Directory)' : '(File)'}`}
        </div>;
    }
}
```
 
The `FileInfoPropertyViewWidgetProvider` is responsible to provide the correct `PropertyViewContentWidget` based on the selection:

```ts
@injectable()
export class FileInfoPropertyViewWidgetProvider extends DefaultPropertyViewWidgetProvider {
 
    readonly id = 'fileinfo';
    readonly label = 'FileInfoPropertyViewWidgetProvider';
 
    private fileInfoWidget: FileInfoPropertyViewWidget;
 
    constructor() {
        super();
        this.fileInfoWidget = new FileInfoPropertyViewWidget();
    }
 
    canHandle(selection: Object | undefined): number {
        return this.isFileSelection(selection) ? 1 : 0;
    }
 
    private isFileSelection(selection: Object | undefined): boolean {
        return !!selection && Array.isArray(selection) && FileSelection.is(selection[0]);
    }
 
    provideWidget(selection: Object | undefined): Promise<FileInfoPropertyViewWidget> {
        return Promise.resolve(this.fileInfoWidget);
    }
 
    updateContentWidget(selection: Object | undefined): void {
        this.getPropertyDataService(selection).then(service => this.fileInfoWidget.updatePropertyViewContent(service, selection));
    }
}
```

In the frontend module of your application you need to bind the `FileInfoPropertyDataService` as well as the `FileInfoPropertyViewWidgetProvider` as follows:
```ts
bind(PropertyDataService).to(FileInfoPropertyDataService).inSingletonScope();
bind(PropertyViewWidgetProvider).to(FileInfoPropertyViewWidgetProvider)
.inSingletonScope();
```
 
Following these few steps should give the reader an idea on how to implement an own property view, consisting of a specific `PropertyViewWidgetProvider` and `PropertyviewDataService`.

#### Outlook

This PR provides the generic capabilities of a global property view that can be extended with specific additional information contributed by specific Theia extensions.
We plan to use this global property view in the context of graphical modeling tools (e.g. diagram editors). A specific use case is, for instance, the implementation of EMF Ecore support for Theia as well as other applications in the context of EMF.cloud.
In EMF.cloud we also have an extension of the property view support of this PR, that allows to dynamically generate editable property views based on JSON schema defined data. If there is interest in the community to have that as part of Theia, we are happy to open a follow-up PR that builds on this PR.

Please see below for an example of a property view (JSON-schema based in this case) in the Theia Ecore editor:

ecore-glsp-property-view

Signed-off-by: Nina Doschek <ndoschek@eclipsesource.com>

### CODE REVIEW

**Review by vince-fugnitto — COMMENTED** [review]
@ndoschek thank you for the contribution, I have not looked nor tested the changes thoroughly yet, but it looks very promising. I did notice a bug however where the `properties` view is not properly cleared when no `editors` are opened:

properties-bug

**Comment by ndoschek** [comment]
Thank you! That's a good point, thanks for noticing! I will look into it tomorrow or Thursday the latest.

**Comment by ndoschek** [comment]
I noticed that widgets opened via the file navigator are not properly activated, which led to missing global selections. This caused the missing update of the property view.
In the method `previewNode` the `WidgetOpenMode` was set to `reveal`. I removed that part to ensure to use the default mode `activate`. Now it is possible for all available preference combinations (`Editor > Enable Preview` as well as `Workbench > List > Open Mode > singleClick | doubleClick`) to ensure correct global selections if widgets are opened via a node in the file navigator.
Please let me know if you see any problems or side-effects with this solution.

File: packages/property-view/src/browser/empty-property-view-widget-provider.tsx
Comment (vince-fugnitto) [inline packages/property-view/src/browser/empty-property-view-widget-provider.tsx:16]: Please add a `newline` between headers and the first import.
Other files in the pull-request will also need to be addressed.

File: packages/property-view/src/browser/property-view-frontend-module.ts
Comment (vince-fugnitto) [inline packages/property-view/src/browser/property-view-frontend-module.ts:19]: nitpick: add this import statement last.

File: packages/property-view/src/browser/style/property-view.css
Comment (vince-fugnitto) [inline packages/property-view/src/browser/style/property-view.css:42]: Ideally we would have variables for these properties.
In addition, we should prefix them by `theia-`: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#css-prefix-global-classes

**Review by vince-fugnitto — COMMENTED** [review]
The functionality works very well :+1: I believe it is a very good initial contribution to the `property-view`.

I have some minor cosmetic changes at the moment, please be sure to rebase the pull-request and resolve the conficts.

File: packages/navigator/src/browser/navigator-model.ts
Comment (ndoschek) [reply packages/navigator/src/browser/navigator-model.ts:97]: I took a deeper look at the behaviour of the file/editor selection itself.
Depending on the preference settings, the editor can either be revealed or opened (and therefore automatically activated).
Regarding the `revealed` mode (which is currently the case for the preferences `Editor > Enable Preview > true` as well as `Workbench > List > Open Mode > singleClick`):
If the editor is revealed, the global selection is not updated and does not contain the revealed editor, as the file node in the navigator stays the current selection. (This is also observable in the GIF from your first [comment](https://github.com/eclipse-theia/theia/pull/8655#pullrequestreview-515724579))

Therefore I would say the behaviour is technically correct as the `property-view` listens to the current selection,
but I agree that it might not be very intuitive to the user.

I can think of two possible ways to go on from here:
- Stick to the behavioural change and activate editors if they are opened via `singleClick` or in `preview` to assure that the global selection is up to date.
- Adapt the `EditorWidget` and `EditorPreviewWidget` to set the global selection if they are revealed.

What's your opinion on this? Thanks in advance.

Comment (vince-fugnitto) [reply packages/navigator/src/browser/navigator-model.ts:97]: > I took a deeper look at the behaviour of the file/editor selection itself.
> Depending on the preference settings, the editor can either be revealed or opened (and therefore automatically activated).
> Regarding the `revealed` mode (which is currently the case for the preferences `Editor > Enable Preview > true` as well as `Workbench > List > Open Mode > singleClick`):
> If the editor is revealed, the global selection is not updated and does not contain the revealed editor, as the file node in the navigator stays the current selection. (This is also observable in the GIF from your first [comment](https://github.com/eclipse-theia/theia/pull/8655#pullrequestreview-515724579))
> 
> Therefore I would say the behaviour is technically correct as the `property-view` listens to the current selection,
> but I agree that it might not be very intuitive to the user.
> 
> I can think of two possible ways to go on from here:
> 
> * Stick to the behavioural change and activate editors if they are opened via `singleClick` or in `preview` to assure that the global selection is up to date.

I do not believe we should modify the current `activate` behavior.

> * Adapt the `EditorWidget` and `EditorPreviewWidget` to set the global selection if they are revealed.

This may be the correct behavior, although I'm not sure myself as I have yet to verify if there are any implications when setting the global selection in this case. Would it be possible to rely on the `shell` to determine the current widget and update the `property-view` then? 

- https://eclipse-theia.github.io/theia/docs/next/classes/core.applicationshell-2.html#onDidChangeCurrentWidget

Comment (ndoschek) [reply packages/navigator/src/browser/navigator-model.ts]: I agree with you and reverted the current `activate` behaviour.
The combination of selection and focus is the main point to look at in my opinion. As it depends on the settings, how editors are opened, I think sticking to the global selection and focus without relying on changes regarding the `ApplicationShell` might be the cleanest way for the `property-view` overall.
Please see my attached screencast, where I show the differences between activated editors and 'only revealed' editor. If editors are revealed, the focus stays in the file navigator widget, until the user decides to activate an editor (or any other widget). This would mean that the `property-view` would depend only on the global selection/focus behaviour and would align without adaptions if there will be changes in the future on that matter.

global-selection-focus

File: CHANGELOG.md
Comment (vince-fugnitto) [inline CHANGELOG.md:38]: minor: the entry will need to be moved under the next release, being `1.10.0`.

**Review by vince-fugnitto — COMMENTED** [review]
@ndoschek the changes are very well implemented and work very well, thank you for updating the pull-request :+1: Given that the extension is currently not pulled by any other `@theia` extension it should be fine to merge and update the `property-view` functionality iteratively without any risk.

I did have minor comments however if you'd like to take a look:

We may want to think about implementing `onActivateRequest` for the widget, since currently when it is opened a warning is logged (since the default implementation is `no-op`). For such a case we can activate the first node in the property-view if you believe it is the proper behavior:

image

File: CHANGELOG.md
Comment (ndoschek) [reply CHANGELOG.md]: Oh of course, my bad! I moved this line up  to a new section

**Comment by vince-fugnitto** [comment]
@ndoschek thank you for the contribution, I waited for others to review as well but I believe it is fine to merge since it does not affect other extensions. Would you mind resolving the conflict and I'll merge right after?

**Comment by ndoschek** [comment]
Thank you very much @vince-fugnitto! I rebased my changes to resolve the conflicts.

**Review by marechal-p — COMMENTED** [review]
LGTM, it worked for me by following the steps.

@eclipse-theia/core is anyone ok with taking in this new extension or should it be in its own repo?

I am not against having it here.

**Comment by marcdumais-work** [comment]
@ndoschek thanks for this contribution. Since SAP is an Eclipse Foundation [member company](https://www.eclipse.org/membership/showMember.php?member_id=665), I think we do not need to register a CQ for this substantial contribution (> 1000 LoC). Can you please look at the official flowchart([eclipse legal process pdf](https://www.eclipse.org/legal/EclipseLegalProcessPoster.pdf)) and confirm - I think this contribution would fit under the `figure 2` case, correct? If so can you also answer the question on `figure 7`? :
```
Confirm that the Contribution:
1. Contains No Cryptography
2. Is Developed from Scratch (without
incorporating content from
elsewhere) – Contact the EMO if
you aren’t sure.
3. Is 100% Project License Code
```

File: packages/property-view/src/browser/property-view-widget.tsx
Comment (ndoschek) [reply packages/property-view/src/browser/property-view-widget.tsx:51]: Thank you for spotting that, I fully agree!

File: packages/property-view/src/browser/empty-property-view-widget-provider.tsx
Comment (ndoschek) [reply packages/property-view/src/browser/empty-property-view-widget-provider.tsx]: @marechal-p Thank you for your review! Fair point, i simplified and stored it in a local component.

**Comment by ndoschek** [comment]
@marcdumais-work I looked at the flowchart and can confirm that `Figure 2` applies, and can answer both `Figure 2` and `Figure 7` with `yes`.

**Comment by marcdumais-work** [comment]
Let me rephrase Paul's question above: 

@eclipse-theia/core is anyone __not ok__ with taking in this new extension, in this repo? If we store it here, it will be published along the other framework extensions, with minimal overhead. If it were in its own repo instead, someone would need to release it from there, separately.

**Comment by marechal-p** [comment]
@ndoschek Do think it would be worthwhile to add some UI tests for it? I don't see much that could be unit-tested, but do you think anything could be tested UI-wise to catch potential regressions? (adding to the `examples/api-tests` test suite).

File: packages/property-view/src/browser/resource-property-view/resource-property-data-service.ts
Comment (ndoschek) [reply packages/property-view/src/browser/resource-property-view/resource-property-data-service.ts:58]: Yes indeed, I returned undefined and refined the check in `resource-property-view-tree-widget.tsx`.

**Review by marechal-p — APPROVED** [review]
LGTM, pinging @eclipse-theia/core again because of the new extension.

File: packages/property-view/README.md
Comment (kittaakos) [inline packages/property-view/README.md:21]: Why is it `monaco` specific?

https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/editor/src/browser/editor.ts#L200

Comment (kittaakos) [reply packages/property-view/README.md:21]: `Navigatable` looks even more generic: https://github.com/eclipse-theia/theia/blob/42042974c59f148974de5e8fd90cba7222cee82f/packages/core/src/browser/navigatable.ts#L29

Comment (ndoschek) [reply packages/property-view/README.md]: You're totally right, thank you for spotting that! :100: 
I make use now of the `Navigatable` interface for the selection check and URI retrieval and could therefore optimize the dependencies and remove `@theia/monaco` completely from this extension. :+1:

**Review by marcdumais-work — DISMISSED** [review]
@ndoschek: Just a quick note to ask that you please squash this PRs commits before eventual merge. In the meantime it's ok to add new ones on top, while making improvements and answering review comments.

**Comment by ndoschek** [comment]
@marcdumais-work All commits in this PR are squashed and rebased to latest master.

**Comment by JonasHelming** [comment]
@marcdumais-work : Was squashing the only remaining change you requested? (that has been done)

**Comment by JonasHelming** [comment]
@ndoschek Could you resolve the conflicts?

**Comment by marcdumais-work** [comment]
@ndoschek is the CI failure something to worry about?

**Comment by ndoschek** [comment]
@marcdumais-work I think the failures were mainly reported because the master branch on my forked repo was not up-to-date and therefore the builds failed, sorry for that!
I disabled the workflow on my fork now, and it looks as the build runs through for this PR.

**Comment by marcdumais-work** [comment]
@marechal-p if you have no further comments, can you merge this contributions?

**Comment by marcdumais-work** [comment]
and/or @vince-fugnitto

File: packages/property-view/README.md
Comment (vince-fugnitto) [inline packages/property-view/README.md:17]: minor: 

```suggestion
The property view widget can be opened/toggled either via menu _View->Properties_ or via shortcut <kbd>Shift+Alt+P</kbd>. It is located in the bottom dock area by default.
```

**Review by vince-fugnitto — COMMENTED** [review]
@ndoschek @JonasHelming some minor updated and it should be good to merge (I'll take care of merging after) :+1: Please be sure to squash after making the commits as well.

File: CHANGELOG.md
Comment (vince-fugnitto) [inline CHANGELOG.md:55]: minor: before merging we should place the following entry under the proper release `1.12.0`. You may also omit the entry, I usually complete the changelog before a release :+1:

**Comment by ndoschek** [comment]
@vince-fugnitto Thank you for your review, I addressed your comments and squashed the commit.

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto Thank you for your review, I addressed your comments and squashed the commit.

@ndoschek looks like something was merged prior to the update which resulted in a conflict. Can you please rebase?

**Review by vince-fugnitto — APPROVED** [review]
Merging! Thank you for your contribution and patience @ndoschek :)

**Pull request merged**
